wrappers/services.go

```diff
@@ -375,120 +376,202 @@ func userDaemonReload() error {
+	preseeding := false
 
 	// note, sysd is not used when preseeding
 	sysd := systemd.New(systemd.SystemMode, inter)
-	var written []string
-	var writtenSystem, writtenUser bool
+	modifiedUnits := make(map[string]*osutil.FileState)
+	var modifiedSystem, modifiedUser bool
 
 	defer func() {
 		if err == nil {
 			return
 		}
-		for _, s := range written {
-			if e := os.Remove(s); e != nil {
-				inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", s, e))
+		for file, state := range modifiedUnits {
+			if state == nil {
+				// we don't have anything to rollback to, so just remove the
+				// file
+				if e := os.Remove(file); e != nil {
+					inter.Notify(fmt.Sprintf("while trying to remove %s due to previous failure: %v", file, e))
+				}
+			} else {
+				// rollback the file to the previous state
+				if e := osutil.EnsureFileState(file, *state); e != nil {
+					inter.Notify(fmt.Sprintf("while trying to rollback %s due to previous failure: %v", file, e))
+				}
 			}
 		}
-		if writtenSystem && !preseeding {
+		if modifiedSystem && !preseeding {
 			if e := sysd.DaemonReload(); e != nil {
 				inter.Notify(fmt.Sprintf("while trying to perform systemd daemon-reload due to previous failure: %v", e))
 			}
 		}
-		if writtenUser && !preseeding {
+		if modifiedUser && !preseeding {
 			if e := userDaemonReload(); e != nil {
 				inter.Notify(fmt.Sprintf("while trying to perform user systemd daemon-reload due to previous failure: %v", e))
 			}
 		}
 	}()
 
-	// create services first; this doesn't trigger systemd
-	for _, app := range s.Apps {
-		if !app.IsService() {
-			continue
+	for s, opts := range snaps {
+		if s.Type() == snap.TypeSnapd {
+			return nil, fmt.Errorf("internal error: adding explicit services for snapd snap is unexpected")
 		}
-		// Generate service file
-		content, err := generateSnapServiceFile(app, opts)
-		if err != nil {
-			return err
-		}
-		svcFilePath := app.ServiceFile()
-		os.MkdirAll(filepath.Dir(svcFilePath), 0755)
-		if err := osutil.AtomicWriteFile(svcFilePath, content, 0644, 0); err != nil {
-			return err
-		}
-		written = append(written, svcFilePath)
-		switch app.DaemonScope {
-		case snap.SystemDaemon:
-			writtenSystem = true
-		case snap.UserDaemon:
-			writtenUser = true
+
+		if opts == nil {
+			opts = &AddSnapServicesOptions{}
 		}
 
-		// Generate systemd .socket files if needed
-		var socketFiles *map[string][]byte
-		socketFiles, err = generateSnapSocketFiles(app)
-		if err != nil {
-			return err
+		// TODO: do we still need this?
+		if opts.Preseeding {
+			preseeding = true
 		}
-		for path, content := range *socketFiles {
-			os.MkdirAll(filepath.Dir(path), 0755)
-			if err = osutil.AtomicWriteFile(path, content, 0644, 0); err != nil {
-				return err
+
+		for _, app := range s.Apps {
+			if !app.IsService() {
+				continue
 			}
-			written = append(written, path)
-		}
 
-		if app.Timer != nil {
+			handleFile := func(path string, desiredContent []byte) error {
```

Comment:
most of this should probably be defined outside of the loops or even as its own helper. for sure not so nested here also because afaict is not using app much except at the very end when recording what it did

Reply:
sure, I moved it outside the loops and made the snap and app arguments to the lambda function